daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/AnfTest.scala

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.speedy
+
+import org.scalatest.{Assertion, WordSpec, Matchers}
+
+import com.daml.lf.speedy.SExpr._
+import com.daml.lf.speedy.SValue._
+import com.daml.lf.speedy.SBuiltin._
+import com.daml.lf.speedy.Anf.flattenToAnf
+import com.daml.lf.speedy.Pretty.SExpr._
+import com.daml.lf.data.Ref._
+
+class AnfTest extends WordSpec with Matchers {
+
+  "identity: [\\x. x]" should {
+    "be transformed to ANF as expected" in {
+      val original = lam(1, arg0)
+      val expected = original
+      testTransform(original, expected)
+    }
+  }
+
+  "twice: [\\f x. f (f x)]" should {
+    "be transformed to ANF as expected" in {
+      val original = lam(2, app(arg0, app(arg0, arg1)))
+      val expected = lam(2, let1(appa(arg0, arg1), appa(arg0, stack1)))
+      testTransform(original, expected)
+    }
+  }
+
+  "thrice: [\\f x. f (f (f x))]" should {
+    "be transformed to ANF as expected" in {
+      val original = lam(2, app(arg0, app(arg0, app(arg0, arg1))))
+      val expected =
+        lam(2, let1(appa(arg0, arg1), let1(appa(arg0, stack1), appa(arg0, stack1))))
+      testTransform(original, expected)
+    }
+  }
+
+  "arithmetic non-atomic: [\\f x. f (x+1)]" should {
+    "be transformed to ANF as expected" in {
+      val original = lam(2, app(arg0, binop(SBAddInt64, arg1, num1)))
+      val expected = lam(2, let1b2(SBAddInt64, arg1, num1, appa(arg0, stack1)))
+      testTransform(original, expected)
+    }
+  }
+
+  "nested (4x non-atomic): [\\f x. f(x+1) - f(x+2)]" should {
+    "be transformed to ANF as expected" in {
+      val original =
+        lam(
+          2,
+          binop(
+            SBSubInt64,
+            app(arg0, binop(SBAddInt64, arg1, num1)),
+            app(arg0, binop(SBAddInt64, arg1, num2))))
+      val expected =
+        lam(
+          2,
+          let1b2(
+            SBAddInt64,
+            arg1,
+            num1,
+            let1(
+              appa(arg0, stack1),
+              let1b2(
+                SBAddInt64,
+                arg1,
+                num2,
+                let1(appa(arg0, stack1), binopa(SBSubInt64, stack3, stack1))))))
+      testTransform(original, expected)
+    }
+  }
+
+  "builtin multi-arg fun: [\\g. (g 1) - (g 2)]" should {
+    "be transformed to ANF as expected" in {
+      val original =
+        lam(1, binop(SBSubInt64, app(arg1, num1), app(arg1, num2)))
+      val expected =
+        lam(1, let1(appa(arg1, num1), let1(appa(arg1, num2), binopa(SBSubInt64, stack2, stack1))))
+      testTransform(original, expected)
+    }
+  }
+
+  "unknown multi-arg fun: [\\f g. f (g 1) (g 2)]" should {
```

Comment:
This test says we're transforming
```
f (g 1) g(2)
```
into
```
let x = g 1 in
let y = f x in
let z = g 2 in
y z
```
doesn't it? To me this looks exactly like the cascade of `let`s and single argument applications we want to avoid.

Reply:
Oh sorry. That is very embarrassing. I had the wrong version of the transform switched on.
What is worse. When I switch to the transform we intend, it crashes in just the same was as attempt #2.

Reply:
I need to look carefully at what is going on here.

Reply:
ok, so the new transform, which avoids the transform to a cascade of single-apps in now switched on
and the AnfTests are updated to show this